### ISSUE TITLE
VPA spec attribute containerPolicies is not honoured correctly

### ISSUE THREAD

**Issue by sachintiptur** [kubernetes/autoscaler#5973]
**Which component are you using?**:
 vertical-pod-autoscaler

**What version of the component are you using?**:

Component version: 0.13.0

**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3", GitCommit:"9e644106593f3f4aa98f8a84b23db5fa378900bd", GitTreeState:"clean", BuildDate:"2023-03-15T13:40:17Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.10", GitCommit:"e770bdbb87cccdc2daa790ecd69f40cf4df3cc9d", GitTreeState:"clean", BuildDate:"2023-05-17T14:06:35Z", GoVersion:"go1.19.9", Compiler:"gc", Platform:"linux/amd64"}
</pre></details>

**What environment is this in?**:
Cloud provider

**What did you expect to happen?**:
The VPA is configured to act/observe only on the configured containers under `containerPolicies`

**What happened instead?**:
All the containers in the pod are being reconfigured/observed and recommendations are updated.
In the example output below, `hamster-sidecar` is not configured under `containerPolicies` but still
it's being observed and reconfigured.

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a deployment having a pod with more than one container
2. Specify the low memory possible in the pod resource requests, so that container is `OOMKIlled` and VPA kicks in
3. Create a VPA resource with `containerPolicies` for only one of the container
4. Describe the vpa and pods, and notice that container resource requests and vpa recomendations are 
updated even for non-configured containers of the pod.

**Anything else we need to know?**:
```

1. snippet of hamster pod yaml output

apiVersion: v1
kind: Pod
metadata:
  annotations:
    vpaObservedContainers: hamster, hamster-sidecar
    vpaUpdates: 'Pod resources updated by hamster-vpa: container 0: memory request;
      container 1: cpu request, memory request'
  creationTimestamp: "2023-07-19T12:54:42Z"
  generateName: hamster-98944-
  labels:
    app: hamster
    pod-template-hash: "989446"
  name: hamster-989446977-cp8wq
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: hamster-989446977
    uid: 7479b672-de3b-493b-97db
  resourceVersion: "191023684"
  uid: 54fa77da-2305-437f
spec:
  containers:
  - args:
    - -c
    - while true; do timeout 0.5s yes >/dev/null; sleep 0.5s; done
    command:
    - /bin/sh
    image: registry.k8s.io/ubuntu-slim:0.1
    imagePullPolicy: IfNotPresent
    name: hamster
    resources:
      requests:
        cpu: 100m
        memory: 50Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-4zvwj
      readOnly: true
  - args:
    - -c
    - while true; do timeout 0.5s yes >/dev/null; sleep 0.5s; done
    command:
    - /bin/sh
    image: registry.k8s.io/ubuntu-slim:0.1
    imagePullPolicy: IfNotPresent
    name: hamster-sidecar
    resources:
      requests:
        cpu: 627m
        memory: "13107200"
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-4zvwj
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: kore-dev-node-7f3ab1
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext:
    runAsNonRoot: true
    runAsUser: 65534
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: kube-api-access
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-ca
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace

2. Snippet of describe vpa output

Name:         hamster-vpa
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  autoscaling.k8s.io/v1
Kind:         VerticalPodAutoscaler
Metadata:
  Creation Timestamp:  2023-07-19T12:53:56Z
  Generation:          3
  Managed Fields:
    API Version:  autoscaling.k8s.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:spec:
        .:
        f:resourcePolicy:
          .:
          f:containerPolicies:
        f:targetRef:
          .:
          f:apiVersion:
          f:kind:
          f:name:
    Manager:      kubectl-create
    Operation:    Update
    Time:         2023-07-19T12:53:56Z
    API Version:  autoscaling.k8s.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:conditions:
        f:recommendation:
          .:
          f:containerRecommendations:
    Manager:         recommender
    Operation:       Update
    Time:            2023-07-19T12:55:40Z
  Resource Version:  191024188
  UID:               512bb6a0-02ca-4493-8175-xxx
Spec:
  Resource Policy:
    Container Policies:
      Container Name:  hamster
      Controlled Resources:
        memory
      Controlled Values:  RequestsAndLimits
      Max Allowed:
        Memory:  500Mi
      Min Allowed:
        Memory:  50Mi
  Target Ref:
    API Version:  apps/v1
    Kind:         Deployment
    Name:         hamster
  Update Policy:
    Update Mode:  Auto
Status:
  Conditions:
    Last Transition Time:  2023-07-19T12:54:40Z
    Status:                True
    Type:                  RecommendationProvided
  Recommendation:
    Container Recommendations:
      Container Name:  hamster
      Lower Bound:
        Memory:  50Mi
      Target:
        Memory:  50Mi
      Uncapped Target:
        Memory:  13107200
      Upper Bound:
        Memory:        500Mi
      Container Name:  hamster-sidecar
      Lower Bound:
        Cpu:     96m
        Memory:  13107200
      Target:
        Cpu:     627m
        Memory:  13107200
      Uncapped Target:
        Cpu:     627m
        Memory:  13107200
      Upper Bound:
        Cpu:     918810m
        Memory:  16852177966
Events:          <none>

```

**Comment by voelzmo** [comment]
Hey @sachintiptur thanks for raising this question! My understanding is that you're trying to turn off VPA recommendations for a certain container in your Pod. You correctly identified that there is the property `.spec.resourcePolicy.containerPolicies[]` to set properties for individual containers. Please note that the field `.spec.resourcePolicy.containerPolicies[].mode` defaults to `Auto` if not specified. This is not only true for an explicitly defined ContainerPolicy, but also for Containers which don't have a ContainerPolicy matching their name (in this case, the Container get defaults for all the values in a ContainerPolicy). Therefore, the list of ContainerPolicies does not work additive (meaning: you specify the things you want, everything else is ignored), but it works the other way around: if you don't want something, you have to exclude it.

Long story short: You can achieve what you want by specifying a ContainerPolicy that explicitly sets `.spec.resourcePolicy.containerPolicies[].mode: "off"` for the Container that shouldn't get any recommendations. Please note that you have to quote `"off"`, [because yaml interprets a literal `off` as the boolean `false`](https://yaml.org/type/bool.html) 🙈 

I can see that this behavior isn't very obvious and [this did come up before](https://github.com/kubernetes/autoscaler/issues/5639), so this is definitely a good pointer that we should improve documentation for this use-case.

/remove-kind bug
/kind support

**Comment by sachintiptur** [comment]
> Hey @sachintiptur thanks for raising this question! My understanding is that you're trying to turn off VPA recommendations for a certain container in your Pod.
> 
> You correctly identified that there is the property `.spec.resourcePolicy.containerPolicies[]` to set properties for individual containers. Please note that the field `.spec.resourcePolicy.containerPolicies[].mode` defaults to `Auto` if not specified. This is not only true for an explicitly defined ContainerPolicy, but also for Containers which don't have a ContainerPolicy matching their name (in this case, the Container get defaults for all the values in a ContainerPolicy). Therefore, the list of ContainerPolicies does not work additive (meaning: you specify the things you want, everything else is ignored), but it works the other way around: if you don't want something, you have to exclude it.
> 
> Long story short: You can achieve what you want by specifying a ContainerPolicy that explicitly sets `.spec.resourcePolicy.containerPolicies[].mode: "off"` for the Container that shouldn't get any recommendations. Please note that you have to quote `"off"`, [because yaml interprets a literal `off` as the boolean `false`](https://yaml.org/type/bool.html) 🙈
> 
> I can see that this behavior isn't very obvious and [this did come up before](https://github.com/kubernetes/autoscaler/issues/5639), so this is definitely a good pointer that we should improve documentation for this use-case.
> 
> /remove-kind bug /kind support

Thanks for the response.
Yes I saw that `.spec.resourcePolicy.containerPolicies[].mode` can be used to turn off VPA recommendations, But the comment for `ResourcePolicy` is a bit contradicting or I would say incomplete as it says below,

```
If not specified, the autoscaler computes recommended
	// resources for all containers in the pod, without additional constraints.
```

which can also mean that if `ResourcePolicy` is configured, then it applies to only to those containers.

**Comment by voelzmo** [comment]
I absolutely agree that this can be improved and totally see how people can interpret it in the same way you did! If you have concrete suggestions, I'm happy to review a PR that adjusts the documentation to make it more clear.

I think the original author wanted to express how it works with the sentence that comes before the one you pasted:
> The resource policy may be used to set constraints on the recommendations for individual containers.

So the only thing `resourcePolicy` can do is set constraints, not specify a list of things you want and automatically switching the VPA off for everything else.

**Comment by sachintiptur** [comment]
@voelzmo That would be really appreciated. 
I propose the comment/doc to reflect something like below:
```
// Controls how the autoscaler computes recommended resources.
	// The resource policy may be used to set constraints on the recommendations
	// for individual containers. If the certain containers need to be excluded from getting the VPA recommendations, then
        // it must be disabled explicitly by using .spec.resourcePolicy.containerPolicies[].mode 
	
```

And also I would like to recommend to add/update an example yml file with multiple container configs.

**Comment by voelzmo** [comment]
/close
PR has been mergen, not sure why it isn't auto-closing?
Thanks, @sachintiptur!
